### PR TITLE
Add ECMP calculator tool

### PR DESCRIPTION
**Pull request by liorghub**
This PR must be merged after PR https://github.com/sonic-net/sonic-buildimage/pull/12840

#### Why I did it
Added ECMP calculator tool.

#### How I did it
New files were added.

#### How to verify it
Manual tests performed according to tests chapter in [HLD](https://github.com/liorghub/SONiC/blob/ecmp_calculator_2/doc/ecmp/ecmp_calculator.md#3101-system-tests)
Automated tests will be added by verification. 

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

**Comment by liat-grozovik** [comment]
@liorghub  is this PR ready for review? anything else is missing?

**Comment by liorghub** [comment]
@liat-grozovik No, this PR is not ready for review. Rest of the code is ready and reviewed internally, I will publish it once I will get verification approval.

**Comment by liorghub** [comment]
@liat-grozovik 
PR was updated and now contains all the code, however, it is marked as draft. 
When verification will approve I will remove the draft.

@nazariig
Can you please approve?

**Comment by liorghub** [comment]
@liat-grozovik Compilation is failing since SDK does not contain hash calculator. PR [Add SDK hash calculator debian to SONiC Bullseye](https://github.com/Mellanox/Switch-SDK-drivers/pull/89) should be merged to fix it.

File: platform/mellanox/docker-syncd-mlnx/Dockerfile.j2
Comment (dgsudharsan) [inline platform/mellanox/docker-syncd-mlnx/Dockerfile.j2:76]: Apart from ecmp_calc.py other files are not binaries but libraries.Can we package them accordingly?

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (dgsudharsan) [inline platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:434]: Instead we could do something like inner_header = self.packet['packet_info'].get('inner')  instead of doing try except. get will return None if there is no such key.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (dgsudharsan) [inline platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:374]: Should ethertype, vlan and other L2 fields be validated as well?

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (dgsudharsan) [inline platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:147]: get_ecmp_id* since we will get only one per route

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py
Comment (dgsudharsan) [inline platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py:52]: During cleanup before the application exits can we invoke sx_api_close?

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py:130]: Done.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py:206]: Yes, I updated function documentation.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py:238]: Unused variable type_ipv6 was removed.

File: platform/mellanox/docker-syncd-mlnx/lib/port_utils.py
Comment (dprital) [inline platform/mellanox/docker-syncd-mlnx/lib/port_utils.py:51]: Check if label port start from 1 or 0

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (dprital) [inline platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:356]: you need to delete the file after using it. There is no delete for this file

File: platform/mellanox/docker-syncd-mlnx/Dockerfile.j2
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/Dockerfile.j2:76]: Done.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:434]: I have changed the code to use get().

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:374]: By design, Ethertype, vlan and other L2 fields are being validated for their types only.
Each field is defined in PACKET_SCHEME as a number or as a string and validation is being performed using python3-jsonschema module.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc.py:356]: Temporary files deletion was added.

File: platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/ecmp_calculator/ecmp_calc_sdk.py:52]: I added call to sx_api_close() in destructor.

File: platform/mellanox/docker-syncd-mlnx/lib/port_utils.py
Comment (liorghub) [reply platform/mellanox/docker-syncd-mlnx/lib/port_utils.py:51]: Label port is the port number printed on the front panel minus one.
On ACS-MSN3700C with 32 ports, first label port is 0 and last is 31 (total of 32 as expected).

**Review by dgsudharsan — APPROVED** [review]

**Review by liat-grozovik — APPROVED** [review]

**Pull request merged**
